File: Catalog core/src/test/java/cz/vhromada/catalog/commons/GenreUtils.java
package cz.vhromada.catalog.commons;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertNotNull;

import java.util.ArrayList;
import java.util.List;

import javax.persistence.EntityManager;

import cz.vhromada.catalog.dao.entities.Genre;

/**
 * A class represents utility class for genres.
 *
 * @author Vladimir Hromada
 */
public final class GenreUtils {

    /**
     * Count of genres
     */
    public static final int GENRES_COUNT = 4;

    /**
     * ID
     */
    public static final Integer ID = 1;

    /**
     * Position
     */
    public static final Integer POSITION = 10;

    /**
     * Creates a new instance of GenreUtils.
     */
    private GenreUtils() {
    }

    /**
     * Returns genre.
     *
     * @param id ID
     * @return genre
     */
    public static Genre newGenre(final Integer id) {
        final Genre genre = new Genre();
        updateGenre(genre);
        if (id != null) {
            genre.setId(id);
            genre.setPosition(id - 1);
        }

        return genre;
    }

    /**
     * Updates genre fields.
     *
     * @param genre genre
     */
    public static void updateGenre(final Genre genre) {
        genre.setName("Name");
    }

    /**
     * Returns genres.
     *
     * @return genres
     */
    public static List<Genre> getGenres() {
        final List<Genre> genres = new ArrayList<>();
        for (int i = 0; i < GENRES_COUNT; i++) {
            genres.add(getGenre(i + 1));
        }
        return genres;
    }

    /**
     * Returns genre for index.
     *
     * @param index index
     * @return genre for index
     */
    public static Genre getGenre(final int index) {
        final Genre genre = new Genre();
        genre.setId(index);
        genre.setName("Genre " + index + " name");
        genre.setPosition(index - 1);

        return genre;
    }

    /**
     * Returns genre.
     *
     * @param entityManager entity manager
     * @param id            genre ID
     * @return genre
     */
    public static Genre getGenre(final EntityManager entityManager, final int id) {
        return entityManager.find(Genre.class, id);
    }

    /**
     * Returns genre with updated fields.
     *
     * @param id            genre ID
     * @param entityManager entity manager
     * @return genre with updated fields
     */
    public static Genre updateGenre(final int id, final EntityManager entityManager) {
        final Genre genre = getGenre(entityManager, id);
        updateGenre(genre);
        genre.setPosition(POSITION);

        return genre;
    }

    /**
     * Returns count of genres.
     *
     * @param entityManager entity manager
     * @return count of genres
     */
    public static int getGenresCount(final EntityManager entityManager) {
        return entityManager.createQuery("SELECT COUNT(g.id) FROM Genre g", Long.class).getSingleResult().intValue();
    }

    /**
     * Asserts genres deep equals.
     *
     * @param expected expected genres
     * @param actual   actual genres
     */
    public static void assertGenresDeepEquals(final List<Genre> expected, final List<Genre> actual) {
        assertNotNull(actual);
        assertEquals(expected.size(), actual.size());
        if (!expected.isEmpty()) {
            for (int i = 0; i < expected.size(); i++) {
                assertGenreDeepEquals(expected.get(i), actual.get(i));
            }
        }
    }

    /**
     * Asserts genre deep equals.
     *
     * @param expected expected genre
     * @param actual   actual genre
     */
    public static void assertGenreDeepEquals(final Genre expected, final Genre actual) {
        assertNotNull(actual);
        assertEquals(expected.getId(), actual.getId());
        assertEquals(expected.getName(), actual.getName());
        assertEquals(expected.getPosition(), actual.getPosition());
    }

}

File: catalog-core/src/test/kotlin/com/github/vhromada/catalog/utils/GenreUtils.kt
package com.github.vhromada.catalog.utils

import com.github.vhromada.catalog.entity.Genre
import org.assertj.core.api.Assertions.assertThat
import org.assertj.core.api.SoftAssertions.assertSoftly
import javax.persistence.EntityManager

/**
 * Updates genre fields.
 *
 * @return updated genre
 */
fun com.github.vhromada.catalog.domain.Genre.updated(): com.github.vhromada.catalog.domain.Genre {
    return copy(name = "Name")
}

/**
 * Updates genre fields.
 *
 * @return updated genre
 */
fun Genre.updated(): Genre {
    return copy(name = "Name")
}

/**
 * A class represents utility class for genres.
 *
 * @author Vladimir Hromada
 */
object GenreUtils {

    /**
     * Count of genres
     */
    const val GENRES_COUNT = 4

    /**
     * Position
     */
    const val POSITION = 10

    /**
     * Returns genres.
     *
     * @return genres
     */
    fun getGenres(): List<com.github.vhromada.catalog.domain.Genre> {
        val genres = mutableListOf<com.github.vhromada.catalog.domain.Genre>()
        for (i in 1..GENRES_COUNT) {
            genres.add(getGenreDomain(index = i))
        }

        return genres
    }

    /**
     * Returns genre.
     *
     * @param id ID
     * @return genre
     */
    fun newGenreDomain(id: Int?): com.github.vhromada.catalog.domain.Genre {
        return com.github.vhromada.catalog.domain.Genre(id = id, name = "", position = if (id == null) null else id - 1)
            .updated()
    }

    /**
     * Returns genre.
     *
     * @param id ID
     * @return genre
     */
    fun newGenre(id: Int?): Genre {
        return Genre(id = id, name = "", position = if (id == null) null else id - 1)
            .updated()
    }

    /**
     * Returns genre for index.
     *
     * @param index index
     * @return genre for index
     */
    fun getGenreDomain(index: Int): com.github.vhromada.catalog.domain.Genre {
        return com.github.vhromada.catalog.domain.Genre(id = index, name = "Genre $index name", position = index + 9)
            .fillAudit(audit = AuditUtils.getAudit())
    }

    /**
     * Returns genre for index.
     *
     * @param index index
     * @return genre for index
     */
    fun getGenre(index: Int): Genre {
        return Genre(id = index, name = "Genre $index name", position = index + 9)
    }

    /**
     * Returns genre.
     *
     * @param entityManager entity manager
     * @param id            genre ID
     * @return genre
     */
    fun getGenre(entityManager: EntityManager, id: Int): com.github.vhromada.catalog.domain.Genre? {
        return entityManager.find(com.github.vhromada.catalog.domain.Genre::class.java, id)
    }

    /**
     * Returns genre with updated fields.
     *
     * @param entityManager entity manager
     * @param id            genre ID
     * @return genre with updated fields
     */
    fun updateGenre(entityManager: EntityManager, id: Int): com.github.vhromada.catalog.domain.Genre {
        val genre = getGenre(entityManager = entityManager, id = id)!!
        return genre
            .updated()
            .copy(position = POSITION)
            .fillAudit(audit = genre)
    }

    /**
     * Returns count of genres.
     *
     * @param entityManager entity manager
     * @return count of genres
     */
    @Suppress("JpaQlInspection")
    fun getGenresCount(entityManager: EntityManager): Int {
        return entityManager.createQuery("SELECT COUNT(g.id) FROM Genre g", java.lang.Long::class.java).singleResult.toInt()
    }

    /**
     * Asserts genres deep equals.
     *
     * @param expected expected list of genres
     * @param actual   actual list of genres
     */
    fun assertDomainGenresDeepEquals(expected: List<com.github.vhromada.catalog.domain.Genre>, actual: List<com.github.vhromada.catalog.domain.Genre>) {
        assertThat(expected.size).isEqualTo(actual.size)
        if (expected.isNotEmpty()) {
            for (i in expected.indices) {
                assertGenreDeepEquals(expected = expected[i], actual = actual[i])
            }
        }
    }

    /**
     * Asserts genre deep equals.
     *
     * @param expected expected genre
     * @param actual   actual genre
     */
    fun assertGenreDeepEquals(expected: com.github.vhromada.catalog.domain.Genre, actual: com.github.vhromada.catalog.domain.Genre) {
        assertSoftly {
            it.assertThat(actual.id).isEqualTo(expected.id)
            it.assertThat(actual.name).isEqualTo(expected.name)
            it.assertThat(actual.position).isEqualTo(expected.position)
            AuditUtils.assertAuditDeepEquals(softly = it, expected = expected, actual = actual)
        }
    }

    /**
     * Asserts genres deep equals.
     *
     * @param expected expected list of genres
     * @param actual   actual list of genres
     */
    fun assertGenresDeepEquals(expected: List<Genre>, actual: List<com.github.vhromada.catalog.domain.Genre>) {
        assertThat(expected.size).isEqualTo(actual.size)
        if (expected.isNotEmpty()) {
            for (i in expected.indices) {
                assertGenreDeepEquals(expected = expected[i], actual = actual[i])
            }
        }
    }

    /**
     * Asserts genre deep equals.
     *
     * @param expected expected genre
     * @param actual   actual genre
     */
    fun assertGenreDeepEquals(expected: Genre, actual: com.github.vhromada.catalog.domain.Genre) {
        assertSoftly {
            it.assertThat(actual.id).isEqualTo(expected.id)
            it.assertThat(actual.name).isEqualTo(expected.name)
            it.assertThat(actual.position).isEqualTo(expected.position)
            it.assertThat(actual.createdUser).isNull()
            it.assertThat(actual.createdTime).isNull()
            it.assertThat(actual.updatedUser).isNull()
            it.assertThat(actual.updatedTime).isNull()
        }
    }

    /**
     * Asserts genres deep equals.
     *
     * @param expected expected list of genres
     * @param actual   actual list of genres
     */
    fun assertGenreListDeepEquals(expected: List<com.github.vhromada.catalog.domain.Genre>, actual: List<Genre>) {
        assertThat(expected.size).isEqualTo(actual.size)
        if (expected.isNotEmpty()) {
            for (i in expected.indices) {
                assertGenreDeepEquals(expected = expected[i], actual = actual[i])
            }
        }
    }

    /**
     * Asserts genre deep equals.
     *
     * @param expected expected genre
     * @param actual   actual genre
     */
    fun assertGenreDeepEquals(expected: com.github.vhromada.catalog.domain.Genre, actual: Genre) {
        assertSoftly {
            it.assertThat(actual.id).isEqualTo(expected.id)
            it.assertThat(actual.name).isEqualTo(expected.name)
            it.assertThat(actual.position).isEqualTo(expected.position)
        }
    }

}
